### PR TITLE
Allow host dataset for IVF-PQ

### PR DESCRIPTION
**Pull request by tfeher**
This PR enables building (or extending) an IVF-PQ index using data in host memory.

### CODE REVIEW

**Comment by tfeher** [comment]
Note for reviewer: `ai_wrapper` class is currently just a copy of `cai_wrapper`, and one shall reduce the redundant code in a follow up PR. I have created a separate issue for this: #1115.

File: python/pylibraft/pylibraft/common/ai_wrapper.py [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+import numpy as np
+
+from pylibraft.common import input_validation
+
+
+class ai_wrapper:
```
Comment (cjnolet) [inline python/pylibraft/pylibraft/common/ai_wrapper.py:21]: Could we have the `cai_wrapper` extend the `ai_wrapper`? This way we can define the common methods once and just have the `cai_wrapper` implement the `copy_to_host()` (and maybe the `ai_wrapper` could implement a `copy_to_device()`?)

File: python/pylibraft/pylibraft/neighbors/ivf_pq/ivf_pq.pyx [hunk 2]
```diff
@@ -465,16 +474,23 @@ def extend(Index index, new_vectors, new_indices, handle=None):
         handle = Handle()
     cdef handle_t* handle_ = <handle_t*><size_t>handle.getHandle()
 
-    vecs_cai = cai_wrapper(new_vectors)
+    try:
+        vecs_cai = cai_wrapper(new_vectors)
```
Comment (cjnolet) [inline python/pylibraft/pylibraft/neighbors/ivf_pq/ivf_pq.pyx:478]: We don't need to do so in this PR but I wonder if we should eventually encapsulate this inside a simple factory function that dispatches to the correct wrapper based on whether it's host or device? I can just see this pattern getting repetitive if we end up using it in more algorithms in the future.

**Review by cjnolet — CHANGES_REQUESTED** [review]
Looks great, just a couple small things.

File: python/pylibraft/pylibraft/common/ai_wrapper.py (hunk 1 above)
Comment (tfeher) [reply python/pylibraft/pylibraft/common/ai_wrapper.py:21]: Changed `cai_wrapper` to be a child class of `ai_wrapper`. 

The `copy_to_host` method is only part of  `device_ndarray`, but not `cai_wrapper`. As I understand the `[c]ai_wrapper` classes just provide convenience for inspecting the array properties, but they do not need to provide copy operation for the data.

File: python/pylibraft/pylibraft/neighbors/ivf_pq/ivf_pq.pyx (hunk 2 above)
Comment (tfeher) [reply python/pylibraft/pylibraft/neighbors/ivf_pq/ivf_pq.pyx]: I have added a factory function.

**Review by tfeher — COMMENTED** [review]
Thanks @cjnolet for the comments, I have updated the PR!

File: python/pylibraft/pylibraft/common/ai_wrapper.py (hunk 1 above)
Comment (cjnolet) [reply python/pylibraft/pylibraft/common/ai_wrapper.py:21]: Ah yes, my mistake. I guess then the only difference really is the name of array interface that's being called so even better to have it this way.

**Review by cjnolet — APPROVED** [review]
Changes LGTM

**Pull request merged**